File: SimulationUtils/hs1d/src_matlab/process/boussinesq_resolution/space_discretization.m
classdef space_discretization
% class defining the discretization in space. Discretization has not to be regular (linearly spaced)

    properties(Access=private)
        xmin                    % minimum x coordinates [m]
        xmax                    % maximum x coordinates [m]
        discretization_type     % linear, log, square, custom
        x                       % (Nx X 1) edges of the space discretization: on these x coordinates Q is computed [m]
%         dx                      % (Nx X 1) space steps array [m]
        x_S
        w_resampled
        soil_depth_resampled
        angle_resampled
    end
    
    properties(Access=public)
        Nx                      % number of discretization time steps
        A
        B
        C 
        Omega
        Omega2
    end
    
    methods(Access=public)
        function obj=space_discretization(xmin,xmax,Nx,discretization_type,xcustom)
            if(nargin<4) discretization_type='lin'; end
            if(nargin<5) xcustom=nan; end 
            obj.xmin=xmin;
            obj.xmax=xmax;
            obj.Nx=Nx;
            obj.discretization_type=discretization_type;
            switch obj.discretization_type
                case 'lin'
                    obj.x=obj.xmin:(obj.xmax-obj.xmin)/obj.Nx:obj.xmax;
                case 'log'
                    if(obj.xmin==0) 
                        log_x_min=-2;
                        obj.xmin=10^log_x_min;
                    else
                        log_x_min=log10(obj.xmin);
                    end
                    log_x_max=log10(obj.xmax);
                    obj.x=logspace(log_x_min,log_x_max,obj.Nx+1);
                case 'custom'
                    obj.x=xcustom;
                    obj.Nx=length(obj.x)-1;
                    obj.xmin=min(obj.x);
                    obj.xmax=max(obj.x);
                case 'square'
                    obj.x=sqrt(obj.xmin):(sqrt(obj.xmax)-sqrt(obj.xmin))/obj.Nx:sqrt(obj.xmax);
                    obj.x=(obj.x).^2;
                otherwise
                    fprintf(strcat('no discretization type corresponding to ',obj.discretization_type,'  \n'));
            end
            obj.x=obj.x';
%             obj.dx=obj.compute_dx;
            obj=obj.compute_x_centered;
        end
        
        function x=get_edges_coordinates(obj)
            x=obj.x;
        end
        
        function x_S=get_center_coordinates(obj)
            x_S=obj.x_S;
        end
        
        function xmin=get_xmin(obj)
            xmin=obj.xmin;
        end
        
        function xmax=get_xmax(obj)
            xmax=obj.xmax;
        end
        
        function Nx=get_Nx(obj)
            Nx=obj.Nx;
        end
        
        function discretization_type=get_discretization_type(obj)
            discretization_type=obj.discretization_type;
        end
                
        function obj=resample_hs1D_spatial_variables(obj,x,w,soil_depth,angle)
%             obj.w_resampled=interpn(x,w,obj.x_S);
%             obj.soil_depth_resampled=interpn(x,soil_depth,obj.x_S);
%             obj.angle_resampled=interpn(x,angle,obj.x);
%             smooth_width_function = fit(x, w,  'smoothingspline', 'SmoothingParam', 0.0001);
% #JM to change
            smooth_width_function = fit(x, w,  'smoothingspline', 'SmoothingParam', 0.9);
            obj.w_resampled=smooth_width_function(obj.x_S);
%             smooth_slope_function = fit(x, angle,  'smoothingspline', 'SmoothingParam', 0.0001);
% #JM to change
            smooth_slope_function = fit(x, angle,  'smoothingspline', 'SmoothingParam', 0.9);
            obj.angle_resampled=smooth_slope_function(obj.x);
            obj.soil_depth_resampled=interpn(x,soil_depth,obj.x_S);
        end
        
        function [x_S,w_resampled,soil_depth_resampled,angle_resampled,x]=get_resampled_variables(obj)
            x_S=obj.x_S;
            w_resampled=obj.w_resampled;
            soil_depth_resampled=obj.soil_depth_resampled;
            angle_resampled=obj.angle_resampled;
            x=obj.x;
        end
                
        function obj=set_matrix_properties(obj)
            obj.A=obj.first_derivative_downstream;
            obj.B=obj.first_derivative_upstream;
            obj.Omega=obj.weight_matrix;
            obj.Omega2=obj.weight_matrix_bis;
%             obj.C=obj.first_derivative_centered;
        end
        
    end
    
    % #JM ideally change public to private ...
    methods(Access=public)
        
        % length(x_S)=Nx
        function obj=compute_x_centered(obj)
            obj.x_S=(obj.x(2:end)+obj.x(1:end-1))/2;
        end
        
        % length(dx)=Nx
        function dx=compute_dx(obj)
            dx=obj.x(2:end)-obj.x(1:end-1);
            %             dx=obj.x(2:end)-obj.x(1:end-1);
            %             dx(end+1)=dx(end);
%             xtemp=[obj.x(1)-(obj.x(2)-obj.x(1));obj.x;obj.x(end)+(obj.x(end)-obj.x(end-1))];
%             dx=(xtemp(3:end)-xtemp(1:end-2))/2;
        end
        
        % length(dx_S)=Nx-1
        function dx_S=compute_dx_centered(obj)
            dx_S=obj.x_S(2:end)-obj.x_S(1:end-1);
        end
        
%% stencil matrix (for basic derivation centered, downstream, upstream) + mean matrix (Omega)        
        %        [ 0   ...  0   0 ...  0 ]
        %        [ -1   1   0   0 ...  0 ]
        %   B  = [ 0   -1   1   0 ...  0 ]
        %        [ ...   ...    ...   ...]
        %        [ 0  ... ...  0  -1   1 ]
        %        [ 0   ...  0   0 ...  0 ]
        % size B (Nx+1) x (Nx)
        
        function B=first_derivative_upstream(obj)
            A=obj.first_derivative_downstream;
            A=A(1:end-1,1:end-1);
            A(A>0)=1;
            A(A<0)=-1;
            
            B=[zeros(1,length(obj.x)-1);A;zeros(1,length(obj.x)-1)];
            dx_S=obj.compute_dx_centered;
            dx_S=[0;dx_S;0];
            B=diag(1./dx_S)*B;
            B=sparse(B);
        end
        
        %        [ -1  1   0  0   0  ...  0 ]
        %   A  = [ 0  -1   1  0  0  ...   0 ]
        %        [ ...      ...      ...  0 ]
        %        [ 0    ...   ...  0  -1  1 ]
        % size A= (Nx) x (Nx+1)
        function A=first_derivative_downstream(obj)
            A=-diag(ones(length(obj.x)-1,1));
            A=[A,zeros(length(obj.x)-1,1)];
            Abis=diag(ones(length(obj.x)-1,1));
            Abis=[zeros(length(obj.x)-1,1),Abis];
            A=A+Abis;
            dx=obj.compute_dx;
            A=diag(1./dx)*A;
            A=sparse(A);
        end
        
        %        [ 0   1  0   0  ...  0 ]
        %   C  = [-1   0  1  0  ...   0 ]
        %        [ ...    ...    ...  1 ]
        %        [ 0  ... ...  0  -1  0 ]
        function C=first_derivative_centered(obj)
            C=-diag(ones(length(obj.x)-1,1),-1);
            C=C+diag(ones(length(obj.x)-1,1),1);
            C=diag(1./obj.dx)*C;
            C=sparse(C);
        end
                
        %        [0.5 0.5  0   0 ...  0 ]
        % Omega= [ 0  0.5 0.5  0 ...  0 ]
        %        [ ...    ...    ... 0.5]
        %        [ 0  ... ...  0  0  0.5]
        function Omega=weight_matrix(obj)
            Omega=obj.first_derivative_upstream;
            Omega(Omega>0)=0.5;
            Omega(Omega<0)=0.5;
            Omega=sparse(Omega);
        end
        
        function Omega2=weight_matrix_bis(obj)
            Omega2=obj.first_derivative_upstream;
            Omega2(Omega2<0)=0;
            Omega2=sparse(Omega2);
        end
    end

end

File: SimulationUtils/hs1d/src_python/Hillslope1D/process/simulation/SpaceDiscretization.py
import numpy as np
from scipy.interpolate import UnivariateSpline
import Hs1D as Hs
import copy

class SpaceDiscretization(object):
    """
        Class defining the discretization in space of the hillslope (linear, log, etc....)
        It's detirmined using X minimal and maximal values, soil depth, angle and width
        of the hillslope for each mesh

        @param :
          - xmin : minimal coordinate of the hillslope
          - xmax : maximal coordinate of the hillslope
          - nx : number of cells to define (if x_custom == -1)
          - discretization_type : type of discretization used : 'linear', 'logarithmic',
              'square' or 'custom'
          - x_custom : -1 or vector of coordinates if discretization_type == 'custom'
          - angle : float or vector of values corresponding to the slope
          - w : float or vector of values corresponding to the width
          - soil_depth : float or vector of values corresponding to the thickness of the layer
          - k : hydraulic conductivity
          - f: porosity
          - z_custom : -1 or a vector altitude of each cell (used to compute the slope
               if x_custom is a vector)

        @attributes:
          - xmin : minimal coordinate of the hillslope
          - xmax : maximal coordinate of the hillslope
          - discretization : type of discretization used
          - xcustom : -1 or vector of coordinates if discretization_type == 'custom'
          - b : computation matrix
          - a : computation matrix
          - omega : computation matrix
          - omega2 : computation matrix
          - x_node : coordinates of cells nodes
          - angle_node : slope of cells on nodes
          - w_node : width of cells on nodes
          - N_node : number of cells
          - soil_depth_node : layer thickness on nodes
          - dx_node : distance between two successive nodes
          - x_edges : coordinates of cells edges
          - N_edges : number of cells edges
          - dx_edges : distance between two successive edges
          - Hs : class Hs1D
    """

    def __init__(self, xmin=0, xmax=100, nx=100, discretization_type='linear', x_custom=-1, \
                 angle=0, w=0, soil_depth=0, k=1/3600, f=0.3, z_custom=-1):
        self.xmin = xmin
        self.xmax = xmax
        self.discretization = discretization_type # linear, log, or custom
        self.xcustom = x_custom
        self.x_edges = self.space_discretization()
        self.Hs = Hs.Hs1D(nx, angle, w, soil_depth, k, f, z_custom, self.x_edges)
        self.x_node = self.compute_x_node()
        self.set_matrix_properties()

    def space_discretization(self):

        if self.discretization == 'linear':
            self.x_edges = np.arange(self.xmin, self.xmax, (self.xmax - self.xmin) / self.N_edges)
        elif self.discretization == 'logarithmic':
            if self.xmin == 0:
                log_xmin = -2
                self.xmin = 10**(log_xmin)
            else:
                log_xmin = np.log10(self.xmin)
            log_xmax = np.log10(self.xmax)
            self.x_edges = np.logspace(log_xmin, log_xmax, self.N_edges)
        elif self.discretization == 'custom':
            self.x_edges = self.xcustom
            self.N_edges = len(self.x_edges)
            self.N_nodes = self.N_edges - 1
            self.xmin = min(self.x_edges)
            self.xmax = max(self.x_edges)
        elif self.discretization == 'square':
            self.x_edges = np.arange(self.xmin**0.5,self.xmax**0.5, \
                                     (self.xmax**0.5 - self.xmin**0.5)/self.N_edges)
            self.x_edges = self.x_edges**2
        else:
            print('no discretization type corresponding to ', self.discretization, '\n')

        return self.x_edges

    def resample_hs1D_spatial_variables(self):
        soil_depth = self.Hs.get_soil_depth_edges()
        w = self.Hs.get_w_edges()
        angle = self.Hs.get_angle_edges()
        self.x_node = self.compute_x_node()
         #UnivariateSpline does the same work as fit ('smoothingspline') in matlab
        smooth_width_function = UnivariateSpline(self.x_edges, w, k=4, s=0.9)
        self.w_node = smooth_width_function(self.x_node)
        smooth_slope_function = UnivariateSpline(self.x_edges, angle, s=0.9)
        self.angle_node = smooth_slope_function(self.x_node)
        self.soil_depth_node = np.interp(np.squeeze(self.x_node),np.squeeze(self.x_edges), soil_depth[0, :])
        return self.w_node, self.angle_node, self.soil_depth_node

    def get_angle_node(self):
        self.resample_hs1D_spatial_variables()
        return self.angle_node

    def get_w_node(self):
        self.resample_hs1D_spatial_variables()
        return self.w_node

    def get_soil_depth_node(self):
        self.resample_hs1D_spatial_variables()
        return self.soil_depth_node

    def set_matrix_properties(self):
        self.a = self.first_derivative_downstream()
        self.b = self.first_derivative_upstream()
        self.omega = self.weight_matrix()
        self.omega2 = self.weight_matrix_bis()
        return self.a
        return self.b
        return self.omega
        return self.omega2

    def compute_x_node(self):
        self.x_node = (self.x_edges[1:] + self.x_edges[0:-1])/2
        return self.x_node

    def compute_dx_edges(self):
        # length(dx) = Nx
        self.dx_edges = self.x_edges[1:]-self.x_edges[0:-1]
        return self.dx_edges

    def compute_dx_node(self):
        # length(dxS) = Nx - 1
        self.dx_node = self.x_node[1:]-self.x_node[0:-1]
        return self.dx_node

        ##########################################################################
        ###### stencil matrix(for basic derivation centered, downstream, upstream)
        ###### + mean matrix (Omega)######
        ##########################################################################

    def first_derivative_upstream(self):
        #######################################
        #         [ 0   ...   0   0   ...   0 ]
        #         [-1    1    0   0   ...   0 ]
        #    B =  [ 0  - 1    1   0   ...   0 ]
        #         [... ...   ... ...  ...  ...]
        #         [ 0  ...   ...  0   - 1   1 ]
        #         [ 0  ...    0   0   ...   0 ]
        #size B (Nx + 1) x(Nx)
        #######################################

        self.a = self.first_derivative_downstream()
        a = copy.copy(self.a[:-1, :-1])
        a[a > 0] = 1
        a[a < 0] = -1
        temp = np.zeros((1, (len(self.x_edges)-1)))
        self.b = np.append(temp, a, axis=0) #+ np.zeros(((len(self.x) - 1),1))
        self.b = np.append(self.b, temp, axis=0)
#        #B = [zeros(1, length(obj.x) - 1); A; zeros(1, length(obj.x) - 1)];
        dx_node = self.compute_dx_node(),
        dx_node = np.hstack((0, np.squeeze(dx_node), 0))
        self.b = np.dot(np.diag(1/dx_node), self.b)
        self.b[np.isnan(self.b)] = 0
        # self.B = np.eye(1 / self.dxS) * self.B
        #self.b = csr_matrix(self.b)
        return self.b

    def first_derivative_downstream(self):
        #########################################
        #       [-1   1   0    0    0   ...   0 ]
        # A =   [ 0 - 1   1    0    0   ...   0 ]
        #       [... ... ...  ...  ...  ...  ...]
        #       [ 0  ... ...  ...   0  - 1    1 ]
        #       size A = (Nx) x(Nx + 1)
        #########################################
        self.dx_edges = self.compute_dx_edges()
        self.a = -np.eye(len(self.x_edges)-1)
        temp = np.zeros(((len(self.x_edges)-1), 1))
        self.a = np.hstack((self.a, temp))
        a1 = np.eye(len(self.x_edges) - 1)
        a2 = np.zeros((len(self.x_edges) - 1, 1))
        abis = np.hstack((a2, a1))
        self.a = self.a + abis
        self.a = np.dot(np.diag(np.squeeze(1/self.dx_edges)), self.a)
#        print(np.size(self.a,0),'  ',np.size(self.a,1))
        # self.A = (np.eye(np.ones(((len(self.x)-1,1)))))*self.A
        #self.A = csr_matrix(self.A)

        return self.a

    def first_derivative_centered(self):
        ######################################################
        #           [ 0      1       0    0   ...           0]
        # C =       [ -1     0       1    0   ...           0]
        #           [ ...   ...     ...                     1]
        #           [ 0     ...     ...        0    - 1     0]
        ######################################################

        self.c = -np.diag((np.ones((len(self.x_edges) - 1, 1)), -1))
        self.c = self.c + np.diag(np.ones((len(self.x_edges) - 1, 1)), 1)
        self.c = np.diag(1 / self.dx_edges) * self.c

        return self.c

    def weight_matrix(self):
        self.omega = copy.copy(self.b)
        self.omega[(self.omega > 0)] = 0.5
        self.omega[(self.omega < 0)] = 0.5
        return self.omega

    def weight_matrix_bis(self):
        self.omega2 = copy.copy(self.b)
        self.omega2[self.omega2 < 0] = 0
        return self.omega2

